File: frontend/src/api/leaderboard.ts
import { apiRequest } from "./http";

export type LeaderboardEntry = {
  user_id: string;
  display_name: string | null;
  total_xp: number;
  weekly_xp?: number;
  rank: number;
  level: number;
  tier_key?: string;
  tier_name?: string;
  projected_tier_name?: string;
  energy_points: number;
  is_current_user: boolean;
  is_promotion_zone?: boolean;
  is_demotion_zone?: boolean;
};

export type DailyFocusItem = {
  document_id: string | null;
  title: string;
  progress_current: number;
  progress_total: number;
  progress_text: string;
};

export type LeaderboardViewer = {
  user_id: string;
  display_name: string;
  total_xp: number;
  weekly_xp?: number;
  rank: number;
  level: number;
  tier_key?: string;
  tier_name?: string;
  projected_tier_name?: string;
  energy_points: number;
  is_promotion_zone?: boolean;
  is_demotion_zone?: boolean;
  daily_focus: DailyFocusItem[];
};

export type LeaderboardSnapshot = {
  scope: string;
  limit: number;
  offset: number;
  week_starts_at?: string | null;
  week_ends_at?: string | null;
  promotion_cutoff_rank?: number;
  demotion_count?: number;
  participants_count?: number;
  has_more: boolean;
  entries: LeaderboardEntry[];
  viewer: LeaderboardViewer;
};

export const getLeaderboardSnapshot = async (
  limit = 25,
  offset = 0,
): Promise<LeaderboardSnapshot> => {
  return apiRequest<LeaderboardSnapshot>(
    `/api/v1/leaderboard?limit=${limit}&offset=${offset}`,
  );
};

export const getLeaderboard = async (limit = 50): Promise<LeaderboardEntry[]> => {
  const snapshot = await getLeaderboardSnapshot(limit, 0);
  return snapshot.entries;
};

File: backend/app/schemas/leaderboard.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class LeaderboardEntryResponse(BaseModel):
    user_id: UUID
    display_name: str | None
    total_xp: int
    weekly_xp: int = 0
    rank: int
    level: int
    tier_key: str = "apprentice"
    tier_name: str = "见习学徒"
    projected_tier_name: str = "见习学徒"
    energy_points: int
    is_current_user: bool = False
    is_promotion_zone: bool = False
    is_demotion_zone: bool = False


class DailyFocusItemResponse(BaseModel):
    document_id: UUID | None
    title: str
    progress_current: int
    progress_total: int
    progress_text: str


class LeaderboardViewerSummaryResponse(BaseModel):
    user_id: UUID
    display_name: str
    total_xp: int
    weekly_xp: int = 0
    rank: int
    level: int
    tier_key: str = "apprentice"
    tier_name: str = "见习学徒"
    projected_tier_name: str = "见习学徒"
    energy_points: int
    is_promotion_zone: bool = False
    is_demotion_zone: bool = False
    daily_focus: list[DailyFocusItemResponse]


class LeaderboardListResponse(BaseModel):
    scope: str
    limit: int
    offset: int
    week_starts_at: datetime | None = None
    week_ends_at: datetime | None = None
    promotion_cutoff_rank: int = 5
    demotion_count: int = 4
    participants_count: int = 0
    has_more: bool
    entries: list[LeaderboardEntryResponse]
    viewer: LeaderboardViewerSummaryResponse
